sql/hive-thriftserver/src/main/java/org/apache/hive/service/auth/HiveAuthFactory.java

```diff
@@ -362,4 +371,34 @@ public static void verifyProxyAccess(String realUser, String proxyUser, String i
     }
   }
 
+  public static boolean needUgiLogin(UserGroupInformation ugi, String principal, String keytab) {
+    return null == ugi || !ugi.hasKerberosCredentials() || !ugi.getUserName().equals(principal) ||
+      !keytab.equals(getKeytabFromUgi());
+  }
+
+  private static String getKeytabFromUgi() {
+    Class<?> clz = UserGroupInformation.class;
+    try {
+      synchronized (clz) {
+        Field field = clz.getDeclaredField("keytabFile");
+        field.setAccessible(true);
+        return (String) field.get(null);
+      }
+    } catch (NoSuchFieldException e) {
+      try {
+        synchronized (clz) {
+          // In Hadoop 3 we don't have "keytabFile" field, instead we should use private method
+          // getKeytab().
+          Method method = clz.getDeclaredMethod("getKeytab");
+          method.setAccessible(true);
+          return (String) method.invoke(UserGroupInformation.getCurrentUser());
+        }
+      } catch (Throwable t) {
+        return null;
+      }
+    } catch (Throwable t) {
+      return null;
+    }
```

Comment:
Let us catch specific exceptions for method/field invocation; not throwable.
The code I had initially written was poor in this regard btw, and was catching throwable - was bad form.